File: Top_Interview_Questions/Hard/23. Merge k Sorted Lists/solution4.py
from typing import List, Optional
from functools import cmp_to_key


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if len(lists) == 0:
            return None
        previous, nodes = None, []
        for head in lists:
            while head is not None:
                nodes.append(head)
                head = head.next
        nodes.sort(key=cmp_to_key(lambda a, b: 1 if a.val < b.val else -1 if a.val > b.val else 0))
        for node in nodes:
            node.next = previous
            previous = node
        return previous

File: Top_Interview_Questions/Hard/23. Merge k Sorted Lists/solution4.cpp
#include <algorithm>
#include <vector>
using namespace std;

/**
 * 想法: 先將 lists flatten 後存到另一個陣列，再由大到小排序，最後用左接的方式接起來
 * 時間: O(mlogm), m: n1+n2+...+nk, 空間: O(m)
 */

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr)
    {
    }
    ListNode(int x) : val(x), next(nullptr)
    {
    }
    ListNode(int x, ListNode *next) : val(x), next(next)
    {
    }
};

bool cmp(ListNode *a, ListNode *b)
{
    return a->val > b->val;
}

class Solution
{
  public:
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        if (lists.size() == 0)
        {
            return nullptr;
        }
        ListNode *previous = nullptr;
        vector<ListNode *> nodes;
        for (int i = 0; i < lists.size(); i++)
        {
            while (lists[i] != nullptr)
            {
                nodes.push_back((lists[i]));
                lists[i] = lists[i]->next;
            }
        }
        sort(nodes.begin(), nodes.end(), cmp);
        for (int i = 0; i < nodes.size(); i++)
        {
            nodes[i]->next = previous;
            previous = nodes[i];
        }
        return previous;
    }
};